<launch>
  <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen" />
  
  <!-- Run a Voxel Grid filter -->
  <node pkg="nodelet" type="nodelet" name="voxel_grid" args="load pcl/VoxelGrid pcl_manager" output="screen">
    <remap from="~input" to="/camera/depth/points" />
    <rosparam>
      filter_field_name: z
      filter_limit_min: 1.5
      filter_limit_max: 2.5
      filter_limit_negative: False
      leaf_size: 0.01
    </rosparam>
  </node>
  
  <!-- Estimate surface normals -->
  <node pkg="nodelet" type="nodelet" name="normal_estimation" args="load pcl/NormalEstimation pcl_manager" output="screen">
    <remap from="~input" to="/voxel_grid/output" />
    <rosparam>
      spatial_locator: 0
      k_search: 0
      radius_search: 0.015
    </rosparam>
  </node>
  
  <!-- Run a Planar Segmentation from Normals filter -->
  <node pkg="nodelet" type="nodelet" name="planar_segmentation" args="load pcl/SACSegmentationFromNormals pcl_manager" output="screen">
    <remap from="~input"   to="/voxel_grid/output" />
    <remap from="~normals" to="/normal_estimation/output" />
    <rosparam>
      # -[ Mandatory parameters
      model_type: 11
      distance_threshold: 0.01
      #max_iterations: 1000
      #method_type: 0
      #optimize_coefficients: true
      #normal_distance_weight: 0.1
      #eps_angle: 0.09
    </rosparam>
  </node>
  
  <!-- Then Extract the Inlier Points -->
  <node pkg="nodelet" type="nodelet" name="extract_plane" args="load pcl/ExtractIndices pcl_manager" output="screen">
    <remap from="~input" to="/voxel_grid/output" />
    <remap from="~indices" to="/planar_segmentation/inliers" />
    <rosparam>
       negative: False
    </rosparam>
  </node>

  <!-- Run a cylinder Segmentation filter -->
  <!--
  <node pkg="nodelet" type="nodelet" name="sac_cylinder" args="load pcl/SACSegmentationFromNormals pcl_manager" output="screen">
    <remap from="~input" to="/sac_plane/inliers" />
    <remap from="~normals" to="/cloud_normals" />
    <rosparam>
      model_type: 5
      max_iterations: 10000
      distance_weight: 0.1
      distance_threshold: 0.05
      radius_search: 0.1
    </rosparam>
  </node>
  -->

  <!-- Then Extract the Inlier Points -->
  <!--
  <node pkg="nodelet" type="nodelet" name="extract_cylinder" args="load pcl/ExtractIndices pcl_manager" output="screen">
    <remap from="~input" to="/cloud_filtered" />
    <remap from="/extract_cylinder/indices" to="/sac_cylinder/inliers" />
    <rosparam>
      negative: True
    </rosparam>
  </node>
  -->
  
</launch>

